WcaOnRails/app/models/locale.rb

```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+class Locale < SimpleDelegator
+  SOME_CHARS = '[\s\S]*?'
+  COMMENT_LINE_GROUP = '((?:\s*#.*\n)*)'
+  KEY_MATCHER = "['\"]?%s['\"]?:"
+  HASHTAG = "hash: "
+
+  attr_accessor :locale
+
+  def initialize(locale, is_translation=false)
+    self.locale = locale.to_s
+    filename = Rails.root.join('config', 'locales_test', "#{locale}.yml")
+    file_content = File.read(filename)
+    super(YAML.load(file_content))
+    if is_translation
+      decorate_with_hashes(file_content, self, "")
+    end
+  end
+
+  def compare_to(base)
+    compare_node_resursive(base[base.locale], self[locale], [])
+  end
+
+  private
+
+  def decorate_with_hashes(text, node, prefix)
+    node.each do |key, value|
+      unless is_leaf?(value)
+        text = decorate_with_hashes(text, value, "#{prefix}#{SOME_CHARS}#{KEY_MATCHER%key}")
+      else
+        # We want leaves to have at least "_translated", and maybe later "_hash"
+        node[key] = { "_translated" => value }
+      end
+      #idea for a simple parser:
+      #state: hash + stack keys
+      #iterate through lines
+      #if match a hash -> hash
+      #if match a key
+        #check nesting level => if ==, last_key = new_key
+        #si <, pop last_key, last_key = new_key
+        #si >, push last_key
+        #+ hash defined -> attach hash to key in loaded object
+        #clear hash
+      regexp = Regexp.new "(#{prefix}#{SOME_CHARS})#{COMMENT_LINE_GROUP}\\s*#{KEY_MATCHER%key}"
+      text = text.gsub(regexp) do |match|
+        # $1 is everything before the comment and the key
+        before = $1
+        # $2 is the hash(es) matched, or empty
+        hashes = $2
+        unless hashes.empty?
+          hash_lines = hashes.split('#').map { |l| l.strip!() }
+          if hash_lines
+            hash = hash_lines.select(&method(:line_filter)).map(&method(:line_cleaner))
+            if hash.size() >= 1
```

Comment:
These parentheses are not necessary, you've it in many places, so I mentione it just here.